crowbar_framework/lib/execute.rb

```diff
@@ -0,0 +1,92 @@
+#
+# Copyright 2013-2016, SUSE Linux GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "open3"
+
+class Execute
+  class << self
+    def bash(user, *params)
+      if user.nil?
+        return bash_as(false, *params)
+      elsif user == "root"
+        return bash_as(true, *params)
+      else
+        execution_uid, execution_gid = uid_gid(user)
+        return false unless execution_uid
+
+        Process.fork do
+          Process.egid = execution_gid
+          Process.euid = execution_uid
+          spawn(*params)
+        end
+
+        _pid, status = Process.wait2
+        return status.success?
+      end
+    end
+
+    def ruby(user, *params)
+      if user.nil?
+        return false
+      elsif user == "root"
+        return bash("root", "ruby", "-e", *params)
+      else
+        return ruby_as(user, *params)
+      end
+    end
+
+    protected
+
+    def bash_as(root, *params)
+      stderr = nil
+      wait_thr = nil
+      if root
+        _stdin, _stdout, stderr, wait_thr = Open3.popen3("sudo", *params)
+      else
+        _stdin, _stdout, stderr, wait_thr = Open3.popen3(*params)
+      end
+
+      Rails.logger.warn(stderr.read.strip) unless wait_thr.value.success?
+      wait_thr.value.success?
+    end
+
+    def ruby_as(user, *params)
+      execution_uid, execution_gid = uid_gid(user)
+      return false unless execution_uid
+
+      Process.fork do
+        Process.egid = execution_gid
+        Process.euid = execution_uid
+        eval(*params)
+      end
+
+      _pid, status = Process.wait2
+      status.success?
```

Comment:
I would say not, when I call a method in ruby then I expect in this case true of false.

For example:
```ruby
if Execute.ruby_as("root", "File.new('test.txt', 'w')")
    ....
end
```

This will not work when we return exit codes.